qualtran/bloqs/multiplexers/apply_lth_bloq.py

```diff
@@ -0,0 +1,121 @@
+#  Copyright 2024 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+from functools import cached_property
+from typing import cast, Iterable, Optional, Sequence, Set, Tuple, Union
+
+import cirq
+import numpy as np
+from attrs import field, frozen
+from numpy.typing import NDArray
+
+from qualtran import Bloq, bloq_example, BloqDocSpec, BoundedQUInt, QBit, Register, Side
+from qualtran._infra.gate_with_registers import merge_qubits, SpecializedSingleQubitControlledGate
+from qualtran.bloqs.multiplexers.unary_iteration_bloq import UnaryIterationGate
+from qualtran.resource_counting import BloqCountT, SympySymbolAllocator
+from qualtran.symbolics import ceil, log2, prod
+
+
+@frozen
+class ApplyLthBloq(UnaryIterationGate, SpecializedSingleQubitControlledGate):  # type: ignore[misc]
+    r"""A SELECT operation that executes one of a list of bloqs $U_l$ based on a quantum index:
+
+    $$
+    \mathrm{SELECT} = \sum_{l}|l \rangle \langle l| \otimes U_l
+    $$
+
+    This bloq uses the unary iteration scheme to apply `ops[selection]` controlled on an optional single-bit `control` register.
+
+    Args:
+        ops: List of bloqs. Each bloq must have identical registers that are all THRU.
+        selection_regs: List of selection registers, 1D selection index inferred if not provided.
+        control_val: If provided, a singly controlled gate is constructed.
+
+    Registers:
+        selection: The indices of the bloq in `ops` to execute.
+        control: The control bit if specified above.
+        [user_spec]: The output registers of the bloqs in `ops`.
+
+    References:
+        [Encoding Electronic Spectra in Quantum Circuits with Linear T Complexity](https://arxiv.org/abs/1805.03662).
+        Babbush et. al. (2018). Section III.A. and Figure 7.
+    """
+
+    ops: NDArray[Bloq] = field(  # type: ignore[type-var]
+        converter=lambda x: np.array(x) if isinstance(x, Iterable) else x,
+        eq=lambda d: tuple(d.flat),
+    )
+    selection_regs: Optional[Tuple[Register, ...]] = None
+    control_val: Optional[int] = None
+
+    def __attrs_post_init__(self):
+        if cast(int, prod(self.ops.shape)) <= 1:
```

Comment:
Since `self.ops` is an NDArray, we can just use `np.prod` with `self.ops.shape` and not worry about casting.